README.md
# CORELS
Certifiably Optimal RulE ListS

**CORELS is a custom discrete optimization
technique for building rule lists over a categorical feature space.**
Our algorithm provides the optimal solution, with a certificate of optimality.
By leveraging algorithmic bounds, efficient data structures,
and computational reuse, we achieve several orders of magnitude speedup in time
and a massive reduction of memory consumption.
Our approach produces optimal rule lists on practical
problems in seconds.
This framework is a novel alternative to CART and other decision tree methods.

* Elaine Angelino, Nicholas Larus-Stone, Daniel Alabi, Margo Seltzer, and Cynthia Rudin.
**Learning Certifiably Optimal Rule Lists for Categorical Data**.
[JMLR](http://www.jmlr.org/papers/volume18/17-716/17-716.pdf), 2018.

* Nicholas Larus-Stone, Elaine Angelino, Daniel Alabi, Margo Seltzer, Vassilios Kaxiras, Aditya Saligrama, Cynthia Rudin
**Systems Optimizations for Learning Certifiably Optimal Rule Lists**
[SysML](http://www.sysml.cc/doc/2018/54.pdf), 2018

* Nicholas Larus-Stone. **Learning Certifiably Optimal Rule Lists: A Case For Discrete Optimization in the 21st Century**.
[Senior thesis](https://dash.harvard.edu/handle/1/38811502), 2017.

* Elaine Angelino, Nicholas Larus-Stone, Daniel Alabi, Margo Seltzer, Cynthia Rudin
**Learning certifiably optimal rule lists for categorical data**
[KDD](https://www.kdd.org/kdd2017/papers/view/learning-certifiably-optimal-rule-lists-for-categorical-data), 2017

CORELS is a custom branch-and-bound algorithm for optimizing rule lists.

**Web UI can be found at:** https://corels.eecs.harvard.edu/

## Overview

* [C/C++ dependencies](#cc-dependencies)
* [Sample command](#sample-command)
* [Usage](#usage)
* [Data format](#data-format)
* [Arguments](#arguments)
* [Example dataset](#example-dataset)
* [Example rule list](#example-rule-list)
* [Optimization algorithm and objective](#optimization-algorithm-and-objective)
* [Data structures](#data-structures)
* [Related work](#related-work)

### C/C++ dependencies

* [gmp](https://gmplib.org/) (GNU Multiple Precision Arithmetic Library)
* [mpfr](http://www.mpfr.org/) (GNU MPFR Library for multiple-precision floating-point computations; depends on gmp)
* [libmpc](http://www.multiprecision.org/) (GNU MPC for arbitrarily high precision and correct rounding; depends on gmp and mpfr)

e.g., [install libmpc](http://brewformulas.org/Libmpc) via [homewbrew](https://brew.sh/) (this will also install gmp and mpfr):

    brew install libmpc

### Sample command

Run the following from the `src/` directory.

    make
    ./corels -r 0.015 -c 2 -p 1 ../data/compas_train.out ../data/compas_train.label ../data/compas_train.minor

### Usage

    ./corels [-b] [-n max_num_nodes] [-r regularization] [-v verbosity] -c (1|2|3|4) -p (0|1|2)
             [-f logging_frequency] -a (1|2|3) [-s] [-L latex_out] data.out data.label [data.minor]

### Data format

For examples, see `compas.out` and `compas.label` in `data/`.  Also see `compas.minor` (optional). Note that our data parsing is fragile and errors will occur if the format is not followed exactly.

* The input data files must be space-delimited text.
* Each line contains `N + 1` fields, where `N` is the number of observations, and ends with `\n` (including the last line).
* In each line, the last `N` fields are `0`'s and `1`'s, and encode a bit vector;
the first field has the format `{text-description}`, where the text between the brackets provides a description of the bit vector.
* There can be no spaces in the text descriptions--words should be separated by dashes or underscores.

### Arguments

**[-b]** Breadth-first search (BFS). You must specify a search policy;
use exactly one of `(-b | -c)`.

**[-n max_num_nodes]** Maximum trie (cache) size.
Stop execution if the number of nodes in the trie exceeds this number.
Default value corresponds to `-n 100000`.

**[-r regularization]** Regularization parameter (optional).
The default value corresponds to `-r 0.01` and can be thought of as adding a
penalty equivalent to misclassifying 1% of data when increasing a rule list's
length by one association rule.

**[-v verbosity]** Verbosity.
Default value corresponds to `-v 0`.
If verbosity is at least `10`, then print machine info.
If verbosity is at least `1000`, then also print input antecedents and labels.

**-c (1|2|3|4)** Best-first search policy.
You must specify a search policy; use exactly one of `(-b | -c)`.
We include four different prioritization schemes:

* Use `-c 1` to prioritize by curiosity (see *Section 5.1 Custom scheduling policies* of our paper).
* Use `-c 2` to prioritize by the lower bound.
* Use `-c 3` to prioritize by the objective.
* Use `-c 4` for depth-first search.

**-p (0|1|2)** Symmetry-aware map (optional).

* Use `-p 0` for no symmetry-aware map (default).
* Use `-p 1` for the permutation map.
* Use `-p 2` for the captured vector map.

**[-f logging_frequency]** Logging frequency.  Default value corresponds to `-f 1000`.

**-a (0|1|2)** Exclude a specific algorithm optimization.
Default value corresponds to `-a 0`.

* Use `-a 0` to include the following optimizations.
* Use `-a 1` to exclude the minimum support bounds (see *Section 3.7 Lower bounds on antecedent support* of our paper).
* Use `-a 2` to exclude the lookahead bound (see *Lemma 2 in Section 3.4 Hierarchical objective lower bound*).

**[-s]** Calculate upper bound on remaining search space size (optional).
This adds a small overhead; the default behavior does not perform the calculation.
With `-s`, we dynamically and incrementally calculate `floor(log10(Gamma(Rc, Q)))`,
where `Gamma(Rc, Q)` is the upper bound (see *Theorem 7 in Section 3.6* of our paper).

**[-L latex_out]** Latex output.  Include this flag to generate a latex representation of the output rule list.

**data.out** File path to training data.  See **Data format**, above.

* Encode `M` antecedents as `M` space-delimited lines of text. Each line contains `N + 1` fields.
* The first field has the format `{antecedent-description}`,
where the text between the brackets describes the antecedent, e.g., `{hair-color:brown}`, or `{age:20-25}`.
* The remaining `N` fields indicate whether the antecedent is true or false for the `N` observations.

**data.label** File path to training labels.  See **Data format**, above.

* Encode labels as two space-delimited lines of text.
*The first line starts with a description of the negative class label, e.g., `{label=0}`;
the remaining `N` fields indicate which of the `N` observations have this label.
* The second line contains analogous (redundant) information for the positive class label.

**data.minor** File path to a bit vector to support use of the equivalent points bound (optional, see *Theorem 20 in Section 3.14* of our paper).

### Example dataset

The files in `data/` were generated from [ProPublica's COMPAS recidivism dataset](https://github.com/propublica/compas-analysis),
specifically, the file `compas-scores-two-years.csv`.

We include one training set (`N = 6489`) and one test set (`N = 721`) from a 10-fold cross-validation experiment.
There are four training data files:

* `compas_train.csv` : `7` categorical and integer-valued features and binary class labels extracted from `compas-scores-two-years.csv`

    sex (Male, Female), age, juvenile-felonies, juvenile-misdemeanors, juvenile-crimes,
    priors, current-charge-degree (Misdemeanor, Felony)

* `compas_train-binary.csv` : `14` equivalent binary features and class labels (included for convenience / use with other algorithms)

    sex:Male, age:18-20, age:21-22, age:23-25, age:26-45, age:>45
    juvenile-felonies:>0, juvenile-misdemeanors:>0, juvenile-crimes:>0
    priors:2-3, priors:=0, priors:=1, priors:>3, current-charge-degree:Misdemeanor

* `compas_train.out` : `155` mined antecedents (binary features and length-2 conjunctions of binary features with support in `[0.005, 0.995]`), e.g.,

    {sex:Female}, {age:18-20}, {sex:Male,current-charge-degree:Misdemeanor}, {age:26-45,juvenile-felonies:>0}

* `compas_train.labels` : class labels

    {recidivate-within-two-years:No}, {recidivate-within-two-years:Yes}

* `compas_train.minor` : bit vector used to support the equivalent points bound

The corresponding test data files are: `compas_test.csv`, `compas_test-binary.csv`, `compas_test.out`, and `compas_test.labels`.

### Example rule list

    if (age=23−25) and (priors=2−3) then predict yes
    else if (age = 18 − 20) then predict yes
    else if (sex = male) and (age = 21 − 22) then predict yes
    else if (priors > 3) then predict yes
    else predict no

This **rule list** predicts two-year recidivism for the [ProPublica two-year recidivism dataset](https://github.com/propublica/compas-analysis),
and was found by CORELS.
Its **prefix** corresponds to the first four rules and its **default rule** corresponds to the last rule.

### Optimization algorithm and objective

CORELS is a custom branch-and-bound algorithm that minimizes the following objective
defined for a rule list `d`, training data `x` and corresponding labels `y`:

    R(d, x, y) = misc(d, x, y) + c * length(d).

This objective is a regularized empirical risk that consists of a loss
and a regularization term that penalizes longer rule lists.

* The loss `misc(d, x, y)` measures `d`'s misclassification error.
* The regularization parameter `c >= 0` is a small constant.
* `length(d)` is the number of rules in `d`'s prefix.

Let `p` be `d`'s prefix.  The following objective lower bound drives
our branch-and-bound procedure:

    b(p, x, y) = misc(p, x, y) + c * length(d) <= R(d, x, y)

where `misc(p, x, y)` is the prefix misclassification error
(due to mistakes made by the prefix, but not the default rule).

### Data structures

* A **trie** (prefix tree) functions as a cache and supports incremental computation.
* A **priority queue** supports multiple best-first search policies, as well as both breadth-first and depth-first search.
* A **map** supports symmetry-aware pruning.

### Related work

CORELS builds directly on:

* Hongyu Yang, Cynthia Rudin, and Margo Seltzer.
**Scalable Bayesian Rule Lists**. [arXiv:1602.08610](https://arxiv.org/abs/1602.08610), 2016. [code](https://github.com/Hongyuy/sbrlmod)

* Benjamin Letham, Cynthia Rudin, Tyler McCormick and David Madigan.
**Interpretable Classifiers Using Rules and Bayesian Analysis: Building a Better Stroke Prediction Model**.
*The Annals of Applied Statistics*, 2015, Vol. 9, No. 3, 1350–1371. [pdf](https://users.cs.duke.edu/~cynthia/docs/LethamRuMcMa15.pdf) [code](https://users.cs.duke.edu/~cynthia/code/BRL_supplement_code.zip)

In particular, CORELS uses a library by Yang et al. for efficiently representing and operating on bit vectors.
See the files [src/rule.h](src/rule.h) and [src/rule.c](src/rule.c).


R/ChangeLog
2019-12-08  Dirk Eddelbuettel  <edd@debian.org>

	* DESCRIPTION (Version, Date): Roll minor version

	* src/alloc.h: Make allocation tracking conditional on TRACK_ALLOC
	* src/cache.h: Idem
	* src/pmap.h: Idem
	* src/utils.h: Idem

	* src/Makevars.in: Mention -DTRACK_ALLOC to enable tracking
	* src/Makevars.win: Idem

2019-11-26  Dirk Eddelbuettel  <edd@debian.org>

	* configure.ac: Build-time configuration
	* configure: Idem
	* src/Makevars.in: Idem
	* src/Makevars.win: Updated default settings
	* cleanup: Added simple script

2019-11-25  Dirk Eddelbuettel  <edd@debian.org>

	* src/cache.cpp: Updated (and renamed) from upstream version
	* src/corels.cpp: Idem
	* src/pmap.cpp: Idem
	* src/rulelib.cpp: Idem
	* src/utils.cpp: Idem
	* src/cache.h: Updated upstream version
	* src/pmap.h: Idem
	* src/queue.h: Idem
	* src/rules.h: Idem
	* src/utils.h: Idem
	* src/run.cpp: New upstream file
	* src/run.h: Idem

	* src/wrapper.cpp: Updated interface
	* src/RcppExports.cpp: Idem
	* R/RcppExports.R: Idem
	* man/corels.Rd: Documented changed variable

	* src/Makevars: Default build now with GMP

2019-11-15  Dirk Eddelbuettel  <edd@debian.org>

	* inst/AUTHORS: Added
	* inst/LICENSE: Idem

	* DESCRIPTION (Description): Reference AUTHORS file
	* README.md: Small addition to TODOs

2019-11-03  Dirk Eddelbuettel  <edd@debian.org>

	* src/wrapper.cpp: Complete interface

	* DESCRIPTION: Basic edits

	* README.md: Added

	* src/alloc.h: Added from corels, *.hh renamed to *.h
	* src/cache.h: Idem
	* pmap.h: Idem
	* src/queue.h: Idem
	* src/rule.h: Idem
	* src/utils.h: Idem

	* src/cache.cc: Added from corels
	* src/corels.cc: Idem
	* src/pmap.cc: Idem
	* src/rulelib.c: Idem
	* src/unused/main.cc: Idem
	* src/utils.cc: Idem

        * inst/sample_data/compas_test-binary.csv: Idem
        * inst/sample_data/compas_test.csv: Idem
        * inst/sample_data/compas_test.label: Idem
        * inst/sample_data/compas_test.out: Idem
        * inst/sample_data/compas_train-binary.csv: Idem
        * inst/sample_data/compas_train.csv: Idem
        * inst/sample_data/compas_train.label: Idem
        * inst/sample_data/compas_train.minor: Idem
        * inst/sample_data/compas_train.out: Idem

	* src/wrapper.cpp: Added interface wrapper
	* src/Makevars: Added minimal build system

	* src/RcppExports.cpp: Added generated file
	* R/RcppExports.R: Idem

	* src/rcpp_hello_world.cpp: Removed file added by package helper
	* man/rcpp_hello_world.Rd: Idem
	* Read-and-delete-me: Idem


src/main.cc
#include "queue.hh"
#include <iostream>
#include <stdio.h>
#include <getopt.h>
#include <string>
#include <string.h>
#include "run.hh"

#define BUFSZ  512

int main(int argc, char *argv[]) {
    const char usage[] = "USAGE: %s [-b] "
        "[-n max_num_nodes] [-r regularization] [-v (%s)] "
        "-c (1|2|3|4) -p (0|1|2) [-f logging_frequency] "
        "-a (0|1|2) [-s] [-L latex_out]"
        "data.out data.label [data.minor]\n\n"
        "%s\n";

    extern char *optarg;
    bool run_bfs = false;
    bool run_curiosity = false;
    int curiosity_policy = 0;
    bool latex_out = false;
    bool use_prefix_perm_map = false;
    bool use_captured_sym_map = false;
    std::set<std::string> verbosity;
    bool verr = false;
    int map_type = 0;
    int max_num_nodes = 100000;
    double c = 0.01;
    char ch;
    bool error = false;
    char error_txt[BUFSZ];
    int freq = 1000;
    int ablation = 0;
    bool calculate_size = false;
    char verbstr[32];
    verbstr[0] = '\0';
    /* only parsing happens here */
    while ((ch = getopt(argc, argv, "bsLc:p:v:n:r:f:a:u:")) != -1) {
        switch (ch) {
        case 'b':
            run_bfs = true;
            break;
        case 's':
            calculate_size = true;
            break;
        case 'c':
            run_curiosity = true;
            curiosity_policy = atoi(optarg);
            break;
        case 'L':
            latex_out = true;
            break;
        case 'p':
            map_type = atoi(optarg);
            use_prefix_perm_map = map_type == 1;
            use_captured_sym_map = map_type == 2;
            break;
        case 'v':
            verr = !parse_verbosity(optarg, &verbstr[0], sizeof(verbstr), &verbosity);
            break;
        case 'n':
            max_num_nodes = atoi(optarg);
            break;
        case 'r':
            c = atof(optarg);
            break;
        case 'f':
            freq = atoi(optarg);
            break;
        case 'a':
            ablation = atoi(optarg);
            break;
        default:
            error = true;
            _snprintf(error_txt, BUFSZ, "unknown option: %c", ch);
        }
    }
    if (max_num_nodes < 0) {
        error = true;
        _snprintf(error_txt, BUFSZ, "number of nodes must be positive");
    }
    if (c < 0) {
        error = true;
        _snprintf(error_txt, BUFSZ, "regularization constant must be postitive");
    }
    if (map_type > 2 || map_type < 0) {
        error = true;
        _snprintf(error_txt, BUFSZ, "symmetry-aware map must be (0|1|2)");
    }
    if ((run_bfs + run_curiosity) != 1) {
        error = true;
        _snprintf(error_txt, BUFSZ,
                "you must use exactly one of (-b | -c)");
    }
    if (argc < 2 + optind) {
        error = true;
        _snprintf(error_txt, BUFSZ,
                "you must specify data files for rules and labels");
    }
    if (run_curiosity && !((curiosity_policy >= 1) && (curiosity_policy <= 4))) {
        error = true;
        _snprintf(error_txt, BUFSZ,
                "you must specify a curiosity type (1|2|3|4)");
    }
    if (verr) {
        error = true;
        _snprintf(error_txt, BUFSZ,
                 "verbosity options must be one or more of (%s), separated with commas (i.e. -v progress,samples)", VERBSTR);
    }
    else {
        if (verbosity.count("samples") && !(verbosity.count("rule") || verbosity.count("label") || verbosity.count("minor") || verbosity.count("loud"))) {
            error = true;
            _snprintf(error_txt, BUFSZ,
                     "verbosity 'samples' option must be combined with at least one of (rule|label|minor|samples|progress|loud|silent)");
        }
        if (verbosity.size() > 1 && verbosity.count("silent")) {
            _snprintf(error_txt, BUFSZ,
                     "verbosity 'silent' option must be passed without any additional verbosity parameters");
        }
    }

    if (error) {
        fprintf(stderr, usage, argv[0], VERBSTR, error_txt);
        return 1;
    }

    if (verbosity.size() == 0) {
        verbosity.insert("progress");
        strcpy(verbstr, "progress");
    }

    if (verbosity.count("silent")) {
        verbosity.clear();
        verbstr[0] = '\0';
    }

    std::map<int, std::string> curiosity_map;
    curiosity_map[1] = "curiosity";
    curiosity_map[2] = "curious_lb";
    curiosity_map[3] = "curious_obj";
    curiosity_map[4] = "dfs";

    argc -= optind;
    argv += optind;

    int nrules, nsamples, nlabels, nsamples_label;
    rule_t *rules, *labels;

    if(rules_init(argv[0], &nrules, &nsamples, &rules, 1) != 0) {
        fprintf(stderr, "Failed to load out file from path: %s\n", argv[0]);
        return 1;
    }

    if(rules_init(argv[1], &nlabels, &nsamples_label, &labels, 0) != 0) {
        fprintf(stderr, "Failed to load label file from path: %s\n", argv[1]);
        rules_free(rules, nrules, 1);
        return 1;
    }

    if(nlabels != 2) {
        fprintf(stderr, "nlabels must be equal to 2, got %d\n", nlabels);
        rules_free(rules, nrules, 1);
        rules_free(labels, nlabels, 0);
        return 1;
    }

    if(nsamples_label != nsamples) {
        fprintf(stderr, "nsamples mismatch between out file (%d) and label file (%d)\n", nsamples, nsamples_label);
        rules_free(rules, nrules, 1);
        rules_free(labels, nlabels, 0);
        return 1;
    }

    int nmeta, nsamples_meta;
    // Equivalent points information is precomputed, read in from file, and stored in meta
    rule_t *meta;
    if (argc == 3) {
        if(rules_init(argv[2], &nmeta, &nsamples_meta, &meta, 0) != 0) {
            fprintf(stderr, "Failed to load minor file from path: %s, skipping...\n", argv[2]);
            meta = NULL;
            nmeta = 0;
        }
        else if(nsamples_meta != nsamples) {
            fprintf(stderr, "nsamples mismatch between out file (%d) and minor file (%d), skipping minor file...\n", nsamples, nsamples_meta);
            rules_free(meta, nmeta, 0);
            meta = NULL;
            nmeta = 0;
        }
    }
    else
        meta = NULL;

    char froot[BUFSZ+32];
    char log_fname[BUFSZ+4+32];
    char opt_fname[BUFSZ+8+32];
    const char* pch = strrchr(argv[0], '/');
    _snprintf(froot, BUFSZ, "../logs/for-%s-%s%s-%s-%s-removed=%s-max_num_nodes=%d-c=%.7f-v=%s-f=%d",
            pch ? pch + 1 : "",
            run_bfs ? "bfs" : "",
            run_curiosity ? curiosity_map[curiosity_policy].c_str() : "",
            use_prefix_perm_map ? "with_prefix_perm_map" :
                (use_captured_sym_map ? "with_captured_symmetry_map" : "no_pmap"),
            meta ? "minor" : "no_minor",
            ablation ? ((ablation == 1) ? "support" : "lookahead") : "none",
            max_num_nodes, c, verbstr, freq);
    snprintf(log_fname, BUFSZ+4+32, "%s.txt", froot);
    snprintf(opt_fname, BUFSZ+8+32, "%s-opt.txt", froot);

    int ret = 0;

    PermutationMap* pmap = NULL;
    CacheTree* tree = NULL;
    Queue* queue = NULL;
    double init = 0.0;
    std::set<std::string> run_verbosity;

    if(run_corels_begin(c, &verbstr[0], curiosity_policy, map_type, ablation, calculate_size,
                        nrules, nlabels, nsamples, rules, labels, meta, freq, &log_fname[0],
                        pmap, tree, queue, init, run_verbosity) == 0)
    {
        while(run_corels_loop(max_num_nodes, pmap, tree, queue) == 0) { }

        std::vector<int> rulelist;
        std::vector<int> classes;

        run_corels_end(&rulelist, &classes, 0, latex_out, rules, labels, &opt_fname[0],
                        pmap, tree, queue, init, verbosity);
    } else {
        fprintf(stderr, "Setup failed!\n");
        ret = 2;
    }

    if(meta) {
        rules_free(meta, nmeta, 0);
    }
    rules_free(rules, nrules, 1);
    rules_free(labels, nlabels, 0);

    return ret;
}
